src/coreclr/jit/emitxarch.cpp

```diff
@@ -887,21 +887,30 @@ bool emitter::AreFlagsSetToZeroCmp(regNumber reg, emitAttr opSize, GenCondition
         return false;
     }
 
-    // Only consider if safe
-    //
     if (!emitCanPeepholeLastIns())
     {
+        // Don't consider if not safe
         return false;
     }
 
     instrDesc*  id      = emitLastIns;
     instruction lastIns = id->idIns();
 
-    if (!id->idHasReg1() || (id->idReg1() != reg))
+    if (!id->idIsReg1Write() || (id->idReg1() != reg))
+    {
+        // Don't consider instructions which didn't write a register
+        return false;
+    }
+
+    if (id->idHasMemWrite() || id->idIsReg2Write())
     {
+        // Don't consider instructions which also wrote a mem location or second register
         return false;
     }
```

Comment:
This is more conservative than it needs to be but brings it back inline with what was previously being checked:
```csharp
    switch (fmt)
    {
        case IF_RWR_CNS:
        case IF_RRW_CNS:
        case IF_RRW_SHF:
        case IF_RWR_RRD:
        case IF_RRW_RRD:
        case IF_RWR_MRD:
        case IF_RWR_SRD:
        case IF_RRW_SRD:
        case IF_RWR_ARD:
        case IF_RRW_ARD:
        case IF_RWR:
        case IF_RRD:
        case IF_RRW:
        case IF_RWR_RRD_RRD:
        case IF_RWR_RRD_MRD:
        case IF_RWR_RRD_ARD:
        case IF_RWR_RRD_SRD:
            break;
        default:
            return false;
    }
```

The general issue was that the check missed cases like `xadd` which wrote two registers or a mem location + a register and while that isn't "incorrect" to do, handling it would require more checks to ensure the correct register write impacted the flags.

I opted for the faster/more conservative fix and will log an issue tracking adding the correctly handling for cases like `xadd`